presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/users/UserEditViewModel.kt

```diff
@@ -1,61 +1,93 @@
+import com.exallium.mvvmcleanexample.presentation.utils.bindable
+import com.exallium.mvvmcleanexample.presentation.utils.delegates.EcoDisposableProperty
+import com.exallium.mvvmcleanexample.presentation.utils.distinctObservable
+import com.exallium.mvvmcleanexample.presentation.utils.ecoDisposable
+import com.exallium.mvvmcleanexample.presentation.utils.notifiable.BaseNotifiableObservable
+import com.exallium.mvvmcleanexample.presentation.utils.notifiable.NotifiableObservable
 import com.jakewharton.rxrelay2.BehaviorRelay
 import com.jakewharton.rxrelay2.PublishRelay
-import io.reactivex.Observable
-import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.disposables.Disposable
 
-class UserEditViewModel(userEditModel: UserEditModel,
-                        private val router: UiRouter) {
+class UserEditViewModel(
+        private val userEditModel: UserEditModel,
+        private val router: UiRouter,
+        private val firstNameRelay: BehaviorRelay<String> = BehaviorRelay.create(),
+        private val lastNameRelay: BehaviorRelay<String> = BehaviorRelay.create(),
+        private val savedUserRelay: PublishRelay<User> = PublishRelay.create(),
+        ecoDisposableProperty: EcoDisposableProperty = ecoDisposable(null),
+        notifiableObservable: NotifiableObservable = BaseNotifiableObservable())
+    : NotifiableObservable by notifiableObservable {
 
-    val firstNameError = ObservableField<String>("")
-    val lastNameError = ObservableField<String>("")
+    init {
+        setDelegator(this)
+    }
 
-    private val compositeDisposable = CompositeDisposable()
+    /* symbol meaning:
+    >> : bind direction left to right, future changes are propagated
+    << : bind direction right to left, future changes are propagated
+    >| : bind direction left to right, future changes are not propagated
+    |< : bind direction right to left, future changes are not propagated
+     */
 
-    private val clickRelay = PublishRelay.create<User>()
-    private val firstNameRelay = BehaviorRelay.create<String>()
-    private val lastNameRelay = BehaviorRelay.create<String>()
+    // view |< >> view model
+    var firstName: String by distinctObservable(
+            initialValue = "",
+            onChange = firstNameRelay::accept)
 
-    init {
-        compositeDisposable.add(userEditModel.results(clickRelay, firstNameRelay, lastNameRelay)
-                .subscribe(this::handleResult, this::handleError))
-    }
+    // view |< >> view model
+    var lastName: String by distinctObservable(
+            initialValue = "",
+            onChange = lastNameRelay::accept)
 
-    fun setSaveClicks(obs: Observable<Unit>) {
-        compositeDisposable.add(obs
-                .map { User(firstNameRelay.value, lastNameRelay.value) }
-                .subscribe(clickRelay))
-    }
+    // view << view model
+    @get:Bindable
+    var firstNameError: String by bindable("")
+
```

Comment:
Good question. Quite honestly, it's unclear. There are different ways of using data binding, implementing `Observable` is considered the most flexible, Extending the framework `BaseObservable` is a convenience as there is some boilerplate involved. 

See this section: https://developer.android.com/topic/libraries/data-binding/index.html#observable_objects
This too covers the approaches: https://medium.com/google-developers/android-data-binding-observability-9de4ff3fe038

What I don't like about `ObservableField` and the primitive counterparts is that they expose public setters on the properties of the view model that are only intended to be observed by the view, where the state is managed internally by the view model.

This is something I forgot to address in this PR but can be easily changed with:
```
@get:Bindable
var firstNameError: String by bindable("")
   private set 
```